include/ensmallen_bits/adam/adam.hpp

```diff
@@ -62,8 +63,9 @@ namespace ens {
  * types included with this distribution or on the ensmallen website.
  *
  * @tparam UpdateRule Adam optimizer update rule to be used.
+ * @tparam DecayPolicy The StepSize decay policy to be used.
  */
-template<typename UpdateRule = AdamUpdate>
+template<typename UpdateRule = AdamUpdate, typename DecayPolicyType = NoDecay>
```

Comment:
Why are we adding `DecayPolicyType` if it isn't ever set to anything other than `NoDecay`?  I can find no uses of that in the codebase.

Reply:
So that if someone wants to change the decay policy , then its possible. A very small use case is 
```
  ens::AdamType<ens::AdamWUpdate> optim;
  optim.DecayPolicy() = ens::CyclicalDecay(50,2,0.03,0.01);
  optim.Optimize(f, coordinates);
```
this is will fail because the decay policy of optim is set to default NoDecay. The way I intend to fix this is by

```
  ens::AdamType<ens::AdamWUpdate, ens::CyclicalDecay> optim;
  optim.DecayPolicy() = ens::CyclicalDecay(50,2,0.03,0.01);
  optim.Optimize(f, coordinates);
```
This is how I added the methods to change Update and decay Policies , if there is a better method then please correct me.

Reply:
I see, thanks for the clarification.  But note that `AdamType<>` is actually just a wrapper around `SGD<>` that exposes some of the functions of the update policy and other bits specific to Adam.  So if you wanted to do

```
ens::AdamType<ens::AdamWUpdate, ens::CyclicalDecay>
```

then it would actually be better to do

```
ens::SGD<ens::AdamWUpdate, ens::CyclicalDecay>
```

instead.  Now, if you are saying that `ens::AdamType<ens::AdamWUpdate, ens::CyclicalDecay>` is a useful optimizer that we should expect people to regularly use, then we can make a typedef for it.  But the documentation as it currently stands mostly expects people to drop in types that represent optimizers and not play with template types.  You can see that the documentation at http://ensmallen.org/docs.html#standard-sgd says

```
Note that StandardSGD is based on the templated type SGD<UpdatePolicyType, DecayPolicyType> with UpdatePolicyType = VanillaUpdate and DecayPolicyType = NoDecay.
```

which is meant to make people aware that it is built on a template type, but it's also meant to keep the documentation as simple as possible.  "Advanced users" who want to play with different template types can take a look through the source and it *should* be straightforward to figure out how to use with different template types.

Anyway, this is probably more of an answer than you asked for, but I was hoping to also clarify some of the reasoning behind my comment. :)